--- /var/www/drupal6/OriModules/js_theming/js_theming.js	2008-10-22 04:31:08.000000000 +0300
+++ js_theming.js	2010-10-12 10:40:32.000000000 +0300
@@ -63,7 +63,8 @@ Drupal.queryStringEncode = function(quer
       newkey = key;
     }
     
-    if (exclude.indexOf(key) >= 0) {
+    var excludestring = exclude.toString();
+    if (excludestring.indexOf(key) >= 0) {
       continue; // Skip exclude keys
     }
     
